packages/go_router/lib/src/shell_state.dart

```diff
@@ -0,0 +1,192 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import '../go_router.dart';
+import 'misc/stateful_navigation_shell.dart';
+
+/// The snapshot of the current state of a [StatefulShellRoute].
+///
+/// Note that this an immutable class, that represents the snapshot of the state
+/// of a StatefulShellRoute at a given point in time. Therefore, instances of
+/// this object should not be stored, but instead fetched fresh when needed,
+/// using the method [StatefulShellRouteState.of].
+@immutable
+class StatefulShellRouteState {
+  /// Constructs a [StatefulShellRouteState].
+  const StatefulShellRouteState({
```

Comment:
If you merge stateful_navigation_shell.dart, you can use a private constructor `StatefulShellRouteState._` that way you can hide more implementation detail

Reply:
👍